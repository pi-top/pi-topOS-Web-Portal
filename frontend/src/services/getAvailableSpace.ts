import api from "./api";

export default async function getAvailableSpace() {
  const { data } = await api.get<string>(`/available-space`);
  try {
    const availableSpace = parseInt(data, 10);
    if (Number.isNaN(availableSpace)) {
      throw new Error();
    }
    return availableSpace;
  } catch (_) {
    return 0;
  }
}
