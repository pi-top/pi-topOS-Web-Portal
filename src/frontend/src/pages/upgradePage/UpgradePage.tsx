import React, { useState, useEffect } from "react";
import { Line as ProgressBar } from "rc-progress";
import prettyBytes from "pretty-bytes";

import Layout from "../../components/layout/Layout";
import Spinner from "../../components/atoms/spinner/Spinner";

import upgradePage from "../../assets/images/upgrade-page.png";
import styles from "./UpgradePage.module.css";

import { OSUpdaterMessage, OSUpdaterMessageType } from "./UpgradePageContainer"
import NewOsVersionDialogContainer from "./newOsVersionDialog/NewOsVersionDialogContainer";

export enum ErrorMessage {
  GenericError = "There was a problem during system update. Please skip - you should be able to update later.",
}

export enum UpgradePageExplanation {
  Preparing = "Checking the size of update...",
  UpgradePreparedWithDownload = "{size} of new packages need to be installed. This might take {time} minutes.",
  UpgradePreparedWithoutDownload = "Some packages need to be installed. This might take a few minutes.",
  InProgress = "Please sit back and relax - this may take some time...",
  Finish = "Great, system update has been successfully installed!\n\nPlease click the {continueButtonLabel} button to restart the application and {continueButtonAction}.",
  WaitingForServer = "Please wait...",
}


export type Props = {
  onNextClick?: () => void;
  onSkipClick?: () => void;
  onBackClick?: () => void;
  onStartUpgradeClick: () => void;
  isCompleted?: boolean;
  message?: OSUpdaterMessage,
  upgradeIsPrepared: boolean,
  upgradeIsRequired: boolean,
  upgradeIsRunning: boolean,
  upgradeFinished: boolean,
  waitingForServer: boolean,
  downloadSize: number,
  error: boolean,
  requireBurn: boolean,
  shouldBurn: boolean,
};

export default ({
  onSkipClick,
  onBackClick,
  onNextClick,
  onStartUpgradeClick,
  isCompleted,
  message,
  upgradeIsPrepared,
  upgradeIsRequired,
  upgradeIsRunning,
  upgradeFinished,
  downloadSize,
  waitingForServer,
  requireBurn,
  shouldBurn,
  error,
}: Props) => {
  const [isNewOsDialogActive, setIsNewOsDialogActive] = useState(false);

  useEffect(() => {
    setIsNewOsDialogActive(requireBurn || shouldBurn);
  }, [requireBurn, shouldBurn]);

  const errorMessage = error && ErrorMessage.GenericError;

  const getExplanation = () => {
    if (error) {
      return ""
    }
    if (waitingForServer) {
      return UpgradePageExplanation.WaitingForServer;
    }
    if (upgradeFinished) {
      return UpgradePageExplanation.Finish.replace(
        "{continueButtonLabel}",
        continueButtonLabel
      ).replace("{continueButtonAction}", onBackClick? "continue" : "finish");
    }
    if (!upgradeIsRequired) {
      return UpgradePageExplanation.Finish.replace(
        "{continueButtonLabel}",
        continueButtonLabel
      ).replace("{continueButtonAction}", onBackClick? "continue" : "finish");
    }
    if (upgradeIsRunning) {
      return UpgradePageExplanation.InProgress;
    }
    if (upgradeIsPrepared) {
      if (downloadSize) {
        return UpgradePageExplanation.UpgradePreparedWithDownload
          .replace("{size}", prettyBytes(downloadSize))
          .replace("{time}", "a few");
      }
      return UpgradePageExplanation.UpgradePreparedWithoutDownload
    }
    return UpgradePageExplanation.Preparing;
  };

  const continueButtonLabel = upgradeIsRequired ? "Update" : onBackClick? "Next" : "Exit"

  return (
    <>
      <Layout
        banner={{
          src: upgradePage,
          alt: "upgrade-page-banner",
        }}
        prompt={
          <>
            OK, I need to be <span className="green">updated</span>
          </>
        }
        explanation={getExplanation()}
        nextButton={{
          onClick: upgradeIsRequired ? onStartUpgradeClick : onNextClick,
          label: continueButtonLabel,
          disabled: !upgradeIsPrepared || upgradeIsRunning || waitingForServer || error
        }}
        skipButton={{ onClick: onSkipClick }}
        showSkip={onSkipClick !== undefined && (isCompleted || error)}
        showBack={onBackClick !== undefined && !upgradeIsRunning}
        backButton={{
          onClick: onBackClick,
          disabled: upgradeIsRunning
        }}
      >

      <NewOsVersionDialogContainer
        active={isNewOsDialogActive}
        requireBurn={requireBurn}
        shouldBurn={shouldBurn}
        onClose={() => setIsNewOsDialogActive(false)}
      />


        { (waitingForServer || !(upgradeIsPrepared || error)) && (
          <>
            <Spinner size={40} />{" "}
          </>
        )}
        {(message?.type === OSUpdaterMessageType.Upgrade) && !waitingForServer && (
          <div className={styles.progress}>
            <ProgressBar
              percent={message.payload.percent}
              strokeWidth={2}
              strokeColor="#71c0b4"
            />
            <span className={styles.message}>{message.payload.message}</span>
          </div>
        )}

        {errorMessage && <span className={styles.error}>{errorMessage}</span>}
      </Layout>
    </>
  );
};
