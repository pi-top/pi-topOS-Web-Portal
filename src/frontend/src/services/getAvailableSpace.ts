import axios from "axios";

import apiBaseUrl from "./apiBaseUrl";

export default async function getAvailableSpace() {
  const { data } = await axios.get<string>(
    `${apiBaseUrl}/available-space`
  );
  try {
    const availableSpace = parseInt(data, 10);
    if (Number.isNaN(availableSpace)) {
      throw new Error();
    }
    return availableSpace;
  } catch (_) {
    return 0;
  };
}
