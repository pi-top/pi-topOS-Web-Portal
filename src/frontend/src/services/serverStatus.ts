import axios from "axios";

import apiBaseUrl from "./apiBaseUrl";

export default async function serverStatus(args: any) {
  const { timeout } = args;
  const { data } = await axios.get<string>(
    `${apiBaseUrl}/status`,
    { timeout: timeout || 10000 }
  );

  return data;
}
