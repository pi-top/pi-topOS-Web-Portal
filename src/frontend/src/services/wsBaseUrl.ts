const wsBaseUrl = "ws://" + window.location.host + ":80/";

export default wsBaseUrl;
