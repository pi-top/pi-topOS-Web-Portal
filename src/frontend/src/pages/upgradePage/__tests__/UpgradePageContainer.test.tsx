import React from "react";
import {
  render,
  fireEvent,
  RenderResult,
  waitForElement,
  wait,
} from "@testing-library/react";

import { act } from "react-dom/test-utils";
import { Server } from "mock-socket";

import UpgradePageContainer, { Props } from "../UpgradePageContainer";
import querySpinner from "../../../../test/helpers/querySpinner";
import { UpgradePageExplanation, ErrorMessage } from "../UpgradePage";
import Messages from "./data/socketMessages.json";
import getAvailableSpace from "../../../services/getAvailableSpace";
import wsBaseUrl from "../../../services/wsBaseUrl";
import serverStatus from "../../../services/serverStatus";
import restartWebPortalService from "../../../services/restartWebPortalService";
import { createModuleBlock } from "typescript";

jest.mock("../../../services/getAvailableSpace");
jest.mock("../../../services/serverStatus");
jest.mock("../../../services/restartWebPortalService");

const getAvailableSpaceMock = getAvailableSpace as jest.Mock;
const serverStatusMock = serverStatus as jest.Mock;
const restartWebPortalServiceMock = restartWebPortalService as jest.Mock;

type ExtendedRenderResult = RenderResult & {
  waitForPreparation: () => Promise<HTMLElement>;
};

let server: Server;

const createServer = () => {
  if (server) {
    server.close();
  }
  return new Server(`${wsBaseUrl}/os-upgrade`);
};

describe("UpgradePageContainer", () => {
  let defaultProps: Props;
  let mount: (props?: Props) => ExtendedRenderResult;

  beforeEach(async () => {
    // we want to have enough space on the device by default
    getAvailableSpaceMock.mockResolvedValue(
      Messages.Size.payload.size.requiredSpace +
        Messages.Size.payload.size.downloadSize +
        1000
    );
    serverStatusMock.mockResolvedValue("OK");
    restartWebPortalServiceMock.mockResolvedValue("OK");

    server = createServer();
    server.on("connection", (socket) => {
      socket.on("message", (data) => {
        if (data === "prepare") {
          socket.send(JSON.stringify(Messages.PrepareStart));
          socket.send(JSON.stringify(Messages.PrepareFinish));
        }

        if (data === "size") {
          socket.send(JSON.stringify(Messages.Size));
        }

        if (data === "start") {
          socket.send(JSON.stringify(Messages.UpgradeStart));
          socket.send(JSON.stringify(Messages.UpgradeStatus));
        }
      });
    });

    defaultProps = {
      goToNextPage: jest.fn(),
      goToPreviousPage: jest.fn(),
      isCompleted: false,
    };

    mount = (props = defaultProps) => {
      const result = render(<UpgradePageContainer {...props} />);
      return {
        ...result,
        waitForPreparation: () =>
          waitForElement(() =>
            result.getByText(
              UpgradePageExplanation.UpgradePrepared
                .replace("{size}", "100 kB")
                .replace("{time}", "a few")
            )
          ),
      };
    };
  });

  afterEach(() => {
    getAvailableSpaceMock.mockRestore();
    serverStatusMock.mockRestore();
    restartWebPortalServiceMock.mockRestore();
  });

  it("renders the correct banner image", async () => {
    const { queryByAltText, waitForPreparation } = mount();
    await wait();

    expect(queryByAltText("upgrade-page-banner")).toMatchSnapshot();

    await waitForPreparation();
  });

  it("renders prompt correctly", async () => {
    const { container: upgradePage, waitForPreparation } = mount();
    await wait();

    const prompt = upgradePage.querySelector(".prompt");
    expect(prompt).toMatchSnapshot();

    await waitForPreparation();
  });

  it("Upgrade button is disabled if upgrade is not ready", async () => {
    const { getByText, waitForPreparation } = mount();

    expect(getByText("Update")).toHaveProperty("disabled", true);

    await waitForPreparation();
  });

  it("skip button is not present while preparing", async () => {
    const { queryByText, waitForPreparation } = mount();

    expect(queryByText("Skip")).not.toBeInTheDocument();

    await waitForPreparation();
  });

  it("calls goToPreviousPage when back button clicked", async () => {
    const { getByText, waitForPreparation } = mount();
    await waitForPreparation();

    fireEvent.click(getByText("Back"));
    expect(defaultProps.goToPreviousPage).toHaveBeenCalled();
  });

  it("renders the preparing upgrade message", async () => {
    const { queryByText, waitForPreparation } = mount();

    expect(queryByText(UpgradePageExplanation.Preparing)).toBeInTheDocument();

    await waitForPreparation();
  });

  it("renders the preparing upgrade spinner", async () => {
    const { container: upgradePage, waitForPreparation } = mount();

    expect(querySpinner(upgradePage)).toBeInTheDocument();

    await waitForPreparation();
  });

  it("does not render skip button while preparing", async () => {
    const { queryByText, waitForPreparation } = mount();

    expect(queryByText("Skip")).not.toBeInTheDocument();

    await waitForPreparation();
  });

  it("renders the upgrade is prepared message when prepared", async () => {
    const { waitForPreparation } = mount();
    await waitForPreparation();
  });

  it("stops rendering the preparing upgrade spinner when prepared", async () => {
    const { container: upgradePage, waitForPreparation } = mount();

    await waitForPreparation();

    expect(querySpinner(upgradePage)).not.toBeInTheDocument();
  });

  it("Upgrade button is enabled when prepared", async () => {
    const { getByText, waitForPreparation } = mount();

    await waitForPreparation();

    expect(getByText("Update")).toHaveProperty("disabled", false);
  });

  it("starts upgrade when Upgrade button clicked", async () => {
    const { getByText, waitForPreparation } = mount();

    await waitForPreparation();
    fireEvent.click(getByText("Update"));

    await waitForElement(() => getByText(UpgradePageExplanation.InProgress));
  });

  describe("when preparation fails", () => {
    beforeEach(async () => {
      server = createServer();
      server.on("connection", (socket) => {
        socket.on("message", () => {
          socket.send(JSON.stringify(Messages.PrepareError));
        });
      });
    });

    it("renders the error message", async () => {
      const { getByText } = mount();
      await wait();

      await waitForElement(() => getByText(ErrorMessage.GenericError));
    });

    it("doesn't render the spinner", async () => {
      const { container: upgradePage, getByText } = mount();
      await waitForElement(() => getByText(ErrorMessage.GenericError));

      expect(querySpinner(upgradePage)).not.toBeInTheDocument();
    });

    it("doesn't render the explanation message", async () => {
      const { getByText, queryByText } = mount();
      await waitForElement(() => getByText(ErrorMessage.GenericError));

      expect(
        queryByText(UpgradePageExplanation.Preparing)
      ).not.toBeInTheDocument();
    });

    it("skip button exists", async () => {
      const { getByText } = mount();
      await wait();

      await waitForElement(() => getByText("Skip"));
    });

    it("calls goToNextPage when Skip button clicked", async () => {
      const { getByText } = mount();

      await waitForElement(() => getByText("Skip"));
      fireEvent.click(getByText("Skip"));

      expect(defaultProps.goToNextPage).toHaveBeenCalled();
    });
  });

  describe("when the upgrade is being installed", () => {
    it("renders the upgrade is in progress message", async () => {
      const { getByText, waitForPreparation, queryByText } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));
      await waitForElement(() => getByText(UpgradePageExplanation.InProgress));

      await waitForElement(() =>
        queryByText(UpgradePageExplanation.InProgress)
      );
    });

    it("renders progress bar correctly", async () => {
      const { getByText, waitForPreparation, container: upgradePage } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));
      await waitForElement(() => getByText(UpgradePageExplanation.InProgress));

      await waitForElement(() => upgradePage.querySelector(".progress"));
      const progressBar = upgradePage.querySelector(".progress");
      expect(progressBar).toMatchSnapshot();
    });

    it("renders the received message", async () => {
      const { getByText, waitForPreparation, queryByText } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));
      await waitForElement(() => getByText(UpgradePageExplanation.InProgress));

      expect(
        queryByText(Messages.UpgradeStatus.payload.message)
      ).toBeInTheDocument();
    });

    it("Upgrade button is disabled", async () => {
      const { getByText, waitForPreparation } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));
      await waitForElement(() => getByText(UpgradePageExplanation.InProgress));

      await waitForElement(() => getByText("Update"));
      expect(getByText("Update")).toHaveProperty("disabled", true);
    });

    it("Back button is disabled", async () => {
      const { getByText, waitForPreparation } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));
      await waitForElement(() => getByText(UpgradePageExplanation.InProgress));

      await waitForElement(() => getByText("Back"));
      expect(getByText("Back")).toHaveProperty("disabled", true);
    });
  });

  describe("when the upgrade fails", () => {
    beforeEach(async () => {
      server = createServer();
      server.on("connection", (socket) => {
        socket.on("message", (data) => {
          if (data === "prepare") {
            socket.send(JSON.stringify(Messages.PrepareStart));
            socket.send(JSON.stringify(Messages.PrepareFinish));
          }

          if (data === "size") {
            socket.send(JSON.stringify(Messages.Size));
          }

          if (data === "start") {
            socket.send(JSON.stringify(Messages.UpgradeStart));
            socket.send(JSON.stringify(Messages.UpgradeStatus));
            socket.send(JSON.stringify(Messages.UpgradeError));
          }
        });
      });
    });

    it("renders the error message", async () => {
      const { getByText, waitForPreparation } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));

      await waitForElement(() => getByText(ErrorMessage.GenericError));
    });

    it("doesn't render the explanation message", async () => {
      const { getByText, waitForPreparation, queryByText } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));

      expect(
        queryByText(UpgradePageExplanation.Preparing)
      ).not.toBeInTheDocument();
    });

    it("Back button is enabled", async () => {
      const { getByText, waitForPreparation } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));

      await waitForElement(() => getByText(ErrorMessage.GenericError));
      expect(getByText("Back")).toHaveProperty("disabled", false);
    });
  });

  describe("when the upgrade finishes", () => {
    beforeEach(() => {
      server = createServer();
      server.on("connection", (socket) => {
        socket.on("message", (data) => {
          if (data === "prepare") {
            socket.send(JSON.stringify(Messages.PrepareStart));
            socket.send(JSON.stringify(Messages.PrepareFinish));
          }

          if (data === "size") {
            socket.send(JSON.stringify(Messages.Size));
          }

          if (data === "start") {
            socket.send(JSON.stringify(Messages.UpgradeStart));
            socket.send(JSON.stringify(Messages.UpgradeStatus));
            socket.send(JSON.stringify(Messages.UpgradeFinish));
          }
        });
      });
    });
    afterEach(() => {
      jest.useRealTimers();
      restartWebPortalServiceMock.mockRestore();
      serverStatusMock.mockRestore();
    })

    it("renders the upgrade finished message", async () => {
      const { getByText, waitForPreparation } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));
      await Promise.all(UpgradePageExplanation.Finish.split("\n").map(async (text, _): Promise<any> => {
        text && await waitForElement(() => getByText(text))
      }))
    });

    it("renders progress bar correctly", async () => {
      const { getByText, waitForPreparation, container: upgradePage } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));
      await Promise.all(UpgradePageExplanation.Finish.split("\n").map(async (text, _): Promise<any> => {
        text && await waitForElement(() => getByText(text))
      }))

      expect(upgradePage.querySelector(".progress")).toMatchSnapshot();
    });

    it("renders the received message", async () => {
      const { getByText, waitForPreparation } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));
      await Promise.all(UpgradePageExplanation.Finish.split("\n").map(async (text, _): Promise<any> => {
        text && await waitForElement(() => getByText(text))
      }))

      expect(
        getByText(Messages.UpgradeFinish.payload.message)
      ).toBeInTheDocument();
    });

    it("Upgrade button is not present", async () => {
      const { getByText, waitForPreparation, queryByText } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));
      await Promise.all(UpgradePageExplanation.Finish.split("\n").map(async (text, _): Promise<any> => {
        text && await waitForElement(() => getByText(text))
      }))

      expect(queryByText("Update")).not.toBeInTheDocument();
    });

    it("Next button is present", async () => {
      const { getByText, waitForPreparation, queryByText } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));
      await Promise.all(UpgradePageExplanation.Finish.split("\n").map(async (text, _): Promise<any> => {
        text && await waitForElement(() => getByText(text))
      }))

      expect(queryByText("Next")).toBeInTheDocument();
    });

    it("requests to restart the pt-web-portal systemd service", async () => {
      const { getByText, waitForPreparation } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));
      await Promise.all(UpgradePageExplanation.Finish.split("\n").map(async (text, _): Promise<any> => {
        text && await waitForElement(() => getByText(text))
      }))
      jest.useFakeTimers();

      fireEvent.click(getByText("Next"));
      await wait();
      jest.runOnlyPendingTimers();
      expect(restartWebPortalServiceMock).toHaveBeenCalled();
    });

    it("displays an error message when restartWebPortalService fails", async () => {
      restartWebPortalServiceMock.mockRejectedValue(new Error("couldn't restart"));
      const { getByText, waitForPreparation } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));
      await Promise.all(UpgradePageExplanation.Finish.split("\n").map(async (text, _): Promise<any> => {
        text && await waitForElement(() => getByText(text))
      }))
      jest.useFakeTimers();

      fireEvent.click(getByText("Next"));
      await wait();
      jest.runOnlyPendingTimers();
      await waitForElement(() => getByText(ErrorMessage.GenericError));
    });

    it("calls goToNextPage when next button clicked", async () => {
      const { getByText, waitForPreparation } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));
      await Promise.all(UpgradePageExplanation.Finish.split("\n").map(async (text, _): Promise<any> => {
        text && await waitForElement(() => getByText(text))
      }))
      jest.useFakeTimers();

      fireEvent.click(getByText("Next"));
      await wait();
      jest.runOnlyPendingTimers();
      jest.runOnlyPendingTimers();
      await wait();
      expect(defaultProps.goToNextPage).toHaveBeenCalled();
    });

    it("calls onBackClick when back button clicked", async () => {
      const { getByText, waitForPreparation } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));
      await Promise.all(UpgradePageExplanation.Finish.split("\n").map(async (text, _): Promise<any> => {
        text && await waitForElement(() => getByText(text))
      }))

      fireEvent.click(getByText("Back"));
      expect(defaultProps.goToPreviousPage).toHaveBeenCalled();
    });

    describe("when pt-web-portal server is restarting", () => {
      afterEach(() => {
        jest.useRealTimers();
        restartWebPortalServiceMock.mockRestore();
        serverStatusMock.mockRestore();
      })

      it("renders a spinner", async () => {
        const { container: upgradePage, getByText, waitForPreparation } = mount();
        await waitForPreparation();
        fireEvent.click(getByText("Update"));
        await Promise.all(UpgradePageExplanation.Finish.split("\n").map(async (text, _): Promise<any> => {
          text && await waitForElement(() => getByText(text))
        }))
        jest.useFakeTimers();

        fireEvent.click(getByText("Next"));
        await wait();
        jest.runOnlyPendingTimers();
        expect(querySpinner(upgradePage)).toBeInTheDocument();
      });

      it("renders a 'please wait' message", async () => {
        const { getByText, waitForPreparation } = mount();
        await waitForPreparation();
        fireEvent.click(getByText("Update"));
        await Promise.all(UpgradePageExplanation.Finish.split("\n").map(async (text, _): Promise<any> => {
          text && await waitForElement(() => getByText(text))
        }))
        jest.useFakeTimers();

        fireEvent.click(getByText("Next"));
        await wait();
        jest.runOnlyPendingTimers();
        await waitForElement(() => getByText(UpgradePageExplanation.WaitingForServer));
      });

      it("probes backend server to determine if its online", async () => {
        const { getByText, waitForPreparation } = mount();
        await waitForPreparation();
        fireEvent.click(getByText("Update"));
        await Promise.all(UpgradePageExplanation.Finish.split("\n").map(async (text, _): Promise<any> => {
          text && await waitForElement(() => getByText(text))
        }))
        jest.useFakeTimers();

        fireEvent.click(getByText("Next"));
        await wait();
        jest.runOnlyPendingTimers();
        expect(serverStatusMock).toHaveBeenCalledTimes(1);
      });

      it("probes backend server until its online", async () => {
        const { getByText, waitForPreparation } = mount();
        await waitForPreparation();
        fireEvent.click(getByText("Update"));
        await Promise.all(UpgradePageExplanation.Finish.split("\n").map(async (text, _): Promise<any> => {
          text && await waitForElement(() => getByText(text))
        }))
        jest.useFakeTimers();

        serverStatusMock.mockRejectedValue(new Error("I'm offline"));
        fireEvent.click(getByText("Next"));

        // checks twice
        jest.runOnlyPendingTimers();
        expect(serverStatusMock).toBeCalledTimes(1);
        jest.runOnlyPendingTimers();
        expect(serverStatusMock).toHaveBeenCalledTimes(2);
        // server is back online
        serverStatusMock.mockResolvedValue("OK");
        jest.runOnlyPendingTimers();
        expect(serverStatusMock).toHaveBeenCalledTimes(3);
        await wait();
        jest.runOnlyPendingTimers();
        // we don't check again
        expect(serverStatusMock).toHaveBeenCalledTimes(3);
      });
    });
  });

  describe("on upgrade error", () => {
    beforeEach(() => {
      server = createServer();
      server.on("connection", (socket) => {
        socket.on("message", (data) => {
          if (data === "prepare") {
            socket.send(JSON.stringify(Messages.PrepareStart));
            socket.send(JSON.stringify(Messages.PrepareFinish));
          }

          if (data === "size") {
            socket.send(JSON.stringify(Messages.Size));
          }

          if (data === "start") {
            socket.send(JSON.stringify(Messages.UpgradeStart));
            socket.send(JSON.stringify(Messages.UpgradeError));
          }
        });
      });
    });

    it("renders the error message", async () => {
      const { getByText, waitForPreparation } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));

      await waitForElement(() => getByText(ErrorMessage.GenericError));
    });

    it("skip button exists", async () => {
      const { getByText, waitForPreparation } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));
      await waitForElement(() => getByText(ErrorMessage.GenericError));

      await waitForElement(() => getByText("Skip"));
      expect(getByText("Skip")).toHaveProperty("disabled", false);
    });

    it("calls goToNextPage when Skip button clicked", async () => {
      const { getByText, waitForPreparation } = mount();
      await waitForPreparation();
      fireEvent.click(getByText("Update"));
      await waitForElement(() => getByText(ErrorMessage.GenericError));

      await waitForElement(() => getByText("Skip"));
      fireEvent.click(getByText("Skip"));

      expect(defaultProps.goToNextPage).toHaveBeenCalled();
    });
  });

  describe("when there's not enough space left on device", () => {
    beforeEach(async () => {
      getAvailableSpaceMock.mockResolvedValue(
        Messages.Size.payload.size.requiredSpace +
          Messages.Size.payload.size.downloadSize -
          1000
      );
    });

    it("renders the error message", async () => {
      const { getByText } = mount();

      await waitForElement(() => getByText(ErrorMessage.GenericError));
    });
  });

  describe("when unable to get available space", () => {
    beforeEach(async () => {
      getAvailableSpaceMock.mockRejectedValue(
        new Error("nah mate couldn't tell ya")
      );
    });

    it("renders the error message", async () => {
      const { getByText } = mount();

      await waitForElement(() => getByText(ErrorMessage.GenericError));
    });
  });
});
