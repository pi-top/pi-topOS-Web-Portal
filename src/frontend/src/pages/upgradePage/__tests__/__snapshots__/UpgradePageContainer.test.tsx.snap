// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpgradePageContainer renders prompt correctly 1`] = `
<h1
  class="prompt"
>
  OK, I need to be 
  <span
    class="green"
  >
    updated
  </span>
</h1>
`;

exports[`UpgradePageContainer renders the correct banner image 1`] = `
<img
  alt="upgrade-page-banner"
  class="image"
  draggable="false"
  src="upgrade-page.png"
  style="max-height: 100%; max-width: 100%;"
/>
`;

exports[`UpgradePageContainer when the upgrade finishes renders progress bar correctly 1`] = `
<div
  class="progress"
>
  <svg
    class="rc-progress-line "
    preserveAspectRatio="none"
    viewBox="0 0 100 2"
  >
    <path
      class="rc-progress-line-trail"
      d="M 1,1
           L 99,1"
      fill-opacity="0"
      stroke="#D9D9D9"
      stroke-linecap="round"
      stroke-width="1"
    />
    <path
      class="rc-progress-line-path"
      d="M 1,1
           L 99,1"
      fill-opacity="0"
      stroke="#71c0b4"
      stroke-linecap="round"
      stroke-width="2"
      style="stroke-dasharray: 0px, 100px; stroke-dashoffset: -0px; transition: stroke-dashoffset 0.3s ease 0s, stroke-dasharray .3s ease 0s, stroke 0.3s linear; transition-duration: 0s, 0s;"
    />
  </svg>
  <span
    class="message"
  >
    Finished upgrade
  </span>
</div>
`;

exports[`UpgradePageContainer when the upgrade is being installed renders progress bar correctly 1`] = `
<div
  class="progress"
>
  <svg
    class="rc-progress-line "
    preserveAspectRatio="none"
    viewBox="0 0 100 2"
  >
    <path
      class="rc-progress-line-trail"
      d="M 1,1
           L 99,1"
      fill-opacity="0"
      stroke="#D9D9D9"
      stroke-linecap="round"
      stroke-width="1"
    />
    <path
      class="rc-progress-line-path"
      d="M 1,1
           L 99,1"
      fill-opacity="0"
      stroke="#71c0b4"
      stroke-linecap="round"
      stroke-width="2"
      style="stroke-dasharray: 50px, 100px; stroke-dashoffset: -0px; transition: stroke-dashoffset 0.3s ease 0s, stroke-dasharray .3s ease 0s, stroke 0.3s linear; transition-duration: .3s, .3s, .3s, .06s;"
    />
  </svg>
  <span
    class="message"
  >
    dpkg-exec: Running dpkg
  </span>
</div>
`;

exports[`UpgradePageContainer when there are major OS updates available and user 'requireBurn' renders the correct message 1`] = `
<div
  aria-pressed="false"
  class="backdrop"
  data-testid="dialog"
  role="button"
>
  <div
    aria-hidden="false"
    aria-modal="true"
    class="dialog newOsVersionAvailableDialog"
    role="dialog"
  >
    <header
      class="header"
    >
      <h3
        class="messageContainer"
        data-testid="dialog-message"
      >
        <span
          class="dialogTitle"
        >
          This version of pi-topOS is no longer supported
        </span>
      </h3>
    </header>
    <div
      class="content"
    >
      <span
        class="osUpgradeWarning"
      >
        This OS version is out of date and not maintained. Your pi-top will not have the latest security updates and features, even if you fully update the system.
      </span>
      <br />
      <span
        class="osUpgradeWarning"
      >
        For more information, go to 
        <button
          class="unstyled link"
          type="button"
        >
          https://pi-top.com/help/out-of-date
        </button>
      </span>
      <br />
      <span
        class="osUpgradeWarning"
      >
        You can also find instructions for downloading and installing the latest version of pi‑topOS from our website.
      </span>
      <br />
      <div
        class="actions"
      >
        <button
          class="styled"
          type="button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`UpgradePageContainer when there are major OS updates available and user 'shouldBurn' renders the correct message 1`] = `
<div
  aria-pressed="false"
  class="backdrop"
  data-testid="dialog"
  role="button"
>
  <div
    aria-hidden="false"
    aria-modal="true"
    class="dialog newOsVersionAvailableDialog"
    role="dialog"
  >
    <header
      class="header"
    >
      <h3
        class="messageContainer"
        data-testid="dialog-message"
      >
        <span
          class="dialogTitle"
        >
          Major OS update required
        </span>
      </h3>
    </header>
    <div
      class="content"
    >
      <span
        class="osUpgradeWarning"
      >
        There is a major OS update available so the update process will take a while. We recommend reburning the SD card.
      </span>
      <br />
      <span
        class="osUpgradeWarning"
      >
        For more information, go to 
        <button
          class="unstyled link"
          type="button"
        >
          https://pi-top.com/help/out-of-date
        </button>
      </span>
      <br />
      <span
        class="osUpgradeWarning"
      >
        You can also find instructions for downloading and installing the latest version of pi‑topOS from our website.
      </span>
      <br />
      <div
        class="actions"
      >
        <button
          class="styled"
          type="button"
        >
          Close
        </button>
      </div>
    </div>
  </div>
</div>
`;
